Feature/movie

antek-ka opened the pull request:
### PR #5

**Movie Page**

- Add global error boundary in App.vue

- Minor style adjustments + scss variable & mixin additions 

- Add api call for get single movie in helper file

- Update MoviePage view from placeholder to draft version:

MoviePage-1440px:
![movie-1440](https://user-images.githubusercontent.com/110823711/186903044-f166ec7f-22dd-4110-ade1-8d03d55e5463.png)

MoviePage-1024px:
![movie-1024](https://user-images.githubusercontent.com/110823711/186903112-e71dd80d-5d2b-434f-b16b-6f12dcf7a112.png)

MoviePage-375px:
![movie-375](https://user-images.githubusercontent.com/110823711/186903174-c5b0e60f-8b40-49c9-901f-2f6f72ddff69.png)

antek-ka committed:
add draft global error boundary

antek-ka committed:
add single movie api call

antek-ka committed:
update fill color to enable css styling in parent component

antek-ka committed:
update class naming + update label-ellipse icon color for color scheme inheritance

antek-ka committed:
update inputInput mixin + add headingFontFace mixin

antek-ka committed:
add new font-size & box-shadow variables

antek-ka committed:
update MoviePage view from placeholder to draft version

rossinek commented on src/components/global/BaseTag.vue:
```diff
@@ -61,12 +64,16 @@ export default defineComponent({
     align-items: center;
     gap: 8px;
 
-    .tag__label {
```
[just a note] tbh I liked the previous one more 🙈  Keep in mind that your classes does not need to fully reflect template structure especially if you are using scoped styles

antek-ka replied:
Ok, it's just that I came to like the way I can chain on sub element classes with `&__` when I learned that You can do that  :joy: .
I changed the markup a bit in a way we can both be pleased - I think - will commit for re-review :wink:

rossinek commented on src/App.vue:
```diff
@@ -27,21 +40,34 @@ export default defineComponent({
-  </main>
+  <template v-if="error">
+    <div class="error">
+      <p>{{ error }}</p>
+      <br />
+      <p>{{ errorInstance }}</p>
+      <br />
+      <p>{{ errorInfo }}</p>
```
Error handling is mostly for users. Do you think it gives your **users** some value to see `errorInstance`& `errorInfo`? 🤔 

Displaying `error` may have some value, but also it may be just some random JS error that does not have any sense for non-dev users. Maybe it's worth to add also some human-readable message? 😁

antek-ka replied:
That was a bit selfish of me, but it's fixed now - something more user friendly :+1:

rossinek replied:
💚

rossinek commented on src/App.vue:
```diff
@@ -19,6 +26,12 @@ export default defineComponent({
+    this.error = err;
+    this.errorInstance = instance;
+    this.errorInfo = info;
+    return false;
```
I don't think it is a good idea to actually return `false` here. Errors here are not expected and doing so will prevent global handler from getting this error - so e.g. error monitoring tools will not work properly.

This may have sense for some minor components that are expected to fail sometimes.

antek-ka replied:
Ok, thanks for the tip!

rossinek commented on src/views/MoviePage.vue:
```diff
@@ -1,13 +1,151 @@
+  },
+  methods: {
+    async getMovieData() {
+      this.movie = await getMovieDetails(this.id);
```
no error handling? 😁

antek-ka replied:
OMG, I experimented with the logic in the component quite a lot and ofc forgot to set it up properly in the end - so sloppy! :imp:

rossinek commented on src/views/MoviePage.vue:
```diff
@@ -1,13 +1,151 @@
+      return this.movie.genre.name;
+    },
+    movieReleaseYear() {
+      return this.movie.release_date.slice(0, 4);
```
it's worth to consider some more stable date parsing/formatting e.g. using `Intl` or just using `Date`

antek-ka replied:
Good point, this CR's theme seems to be that I rely on API data constancy too much.. :sweat:

rossinek replied:
nice observation 👏

rossinek commented on src/views/MoviePage.vue:
```diff
@@ -1,13 +1,151 @@
+  },
+  data() {
+    return {
+      id: this.$route.params.movieId,
```
It's not the best idea to keep it in the state. It may be tricky to see but: it may change❗ 

Vue reuses components if virtual dom matches and is just passing different props to it. Especially it works like this for route components: if you are on `/movies/ID-1` and navigate to `/movies/ID-2` the same component will be used: e.g. `mounted` will be used only once.

What can we do?

- use `computed` for getting stuff from `$route`
- use watcher instead of using `mounted`/`created` for fetching data
e.g.
```js
watch: {
  movieId: {
    handler(movieId) {
      this.movie = await getMovieDetails(movieId);
    },
    immediate: true, // call on created
  }
}

```

antek-ka replied:
I had to really fight this one out :sweat_smile:
Turns out that the watcher doesn't work if the `immediate: true` option comes after the logic...
Well, now it works and I'm not going to ever forget this.

rossinek replied:
> Turns out that the watcher doesn't work if the immediate: true option comes after the logic...

wut? 🤨
that's definitely not true: https://codesandbox.io/s/keen-ritchie-i6g49v?file=/src/components/HelloWorld.vue

maybe there was something wrong with nesting e.g.
```js
watch: {
  movieId: {
    handler(movieId) {
      this.movie = await getMovieDetails(movieId);
    },
  }
  immediate: true,
}
```
instead of 

```js
watch: {
  movieId: {
    handler(movieId) {
      this.movie = await getMovieDetails(movieId);
    },
    immediate: true,
  }
}
```

antek-ka committed:
move errorCaptured logic to separate component
+ add ErrorBoundary
+ add Suspense wrapper
+ add DataLoader as Suspense fallback

antek-ka committed:
update component layout & class naming

antek-ka committed:
create draft DataLoader component

antek-ka committed:
create draft ErrorBoundary component

antek-ka committed:
move movieId to computed properties
+ move data fetching logic to movieId watcher
+ change release_date parsing method to getFullYear()

rossinek commented on src/App.vue:
```diff
@@ -4,12 +4,23 @@ import { defineComponent } from 'vue';
+    DataLoader,
+    ErrorBoundary
+  },
+  data() {
+    return {
+      error: null,
+      errorInstance: null,
+      errorInfo: ''
+    };
```
not used anymore?

antek-ka replied:
Yup, missed to delete these properties - sorry!

rossinek commented on src/App.vue:
```diff
@@ -27,21 +38,24 @@ export default defineComponent({
-    <router-view class="main-view"></router-view>
-  </main>
+  <ErrorBoundary>
+    <Suspense>
```
nice idea 👏

rossinek commented on src/views/MoviePage.vue:
```diff
@@ -1,13 +1,160 @@
+        try {
+          this.movie = await getMovieDetails(movieId);
+        } catch (error) {
+          console.error(error);
```
hmm.. do you expect users to look at the console? 😁  if you don't have better idea for now and just want to move forward - even `alert` is better than nothing 😉

antek-ka replied:
Yeah, so basically the `console.error()` is more for me just for development purposes - I intend to delete it later.
In fact, I would like to delete it right now - could You just tell me if it's OK if I don't catch the eventual `getMovieDetails()` error here and let it pass to the error boundary?
The effect looks like this:
![image](https://user-images.githubusercontent.com/110823711/187212527-0cdd8491-0f8f-4772-8d19-7cb5d7a4dda6.png)

rossinek replied:
sure thing! It's not ideal but definitely better than nothing :)

rossinek reviewed (approved):
Approved with some minor comments.

Separate component for `ErrorBoundary` is a great idea! 👏
